lib/active_merchant/billing/gateways/cyber_source.rb

```diff
@@ -611,15 +611,25 @@ def add_auth_service(xml, payment_method, options)
 
       def add_normalized_threeds_2_data(xml, payment_method, options)
         threeds_2_options = options[:three_d_secure]
+        cc_brand = card_brand(payment_method).to_sym
 
-        xml.tag!('cavv', threeds_2_options[:cavv]) if threeds_2_options[:cavv] && card_brand(payment_method).to_sym != :master
+        xml.tag!('cavv', threeds_2_options[:cavv]) if threeds_2_options[:cavv] && cc_brand != :master
         xml.tag!('cavvAlgorithm', threeds_2_options[:cavv_algorithm]) if threeds_2_options[:cavv_algorithm]
         xml.tag!('paSpecificationVersion', threeds_2_options[:version]) if threeds_2_options[:version]
         xml.tag!('directoryServerTransactionID', threeds_2_options[:ds_transaction_id]) if threeds_2_options[:ds_transaction_id]
-        xml.tag!('commerceIndicator', options[:commerce_indicator] || ECI_BRAND_MAPPING[card_brand(payment_method).to_sym])
+        xml.tag!('commerceIndicator', options[:commerce_indicator] || ECI_BRAND_MAPPING[cc_brand])
         xml.tag!('eciRaw', threeds_2_options[:eci]) if threeds_2_options[:eci]
 
-        xid = threeds_2_options[:xid] || threeds_2_options[:cavv]
+        xid = case cc_brand
+              when :master
+                # Only send xid for mastercard on 3ds 1.x
+                threeds_2_options[:xid] if threeds_2_options[:version].nil?
```

Comment:
My understanding is:
- 1.x 
   - if `xid` is present, send it
   - do not send the `cavv` in the `xid` XML element
- 2.x 
   - if `xid` is present, send it
   - else if `cavv`  is present, send `cavv` in the `xid` XML element for all card brands except `:master`


Does that match yours? 
If it does, I read the code is doing something else.

Reply:
Oh, I misunderstood the logic then, thanks for the clarification!